### PR TITLE
Simplify XPackLicenseState tests

### PR DESCRIPTION
**Pull request by rjernst**
Tests for the license state historically test each feature at random
license levels. But this is unnecessary because what really matters are
the combinations of current license state and license level of a
feature. This commit removes the dozens of feature specific tests in
favor of simple tests for a feature of each license level, asserting
whether that feature is allowed or not.

### CODE REVIEW

File: x-pack/plugin/core/src/test/java/org/elasticsearch/license/XPackLicenseStateTests.java [hunk 1]
```diff
@@ -201,246 +200,44 @@ public void testMonitoringAckNotBasicToBasic() {
...
+        assertAllowed(BASIC, true, s -> s.isAllowed(feature), false);
+        assertAllowed(STANDARD, true, s -> s.isAllowed(feature), false);
+        assertAllowed(GOLD, true, s -> s.isAllowed(feature), false);
+        assertAllowed(PLATINUM, true, s -> s.isAllowed(feature), true);
+        assertAllowed(ENTERPRISE, true, s -> s.isAllowed(feature), true);
+        assertAllowed(TRIAL, true, s -> s.isAllowed(feature), true);
+    }
+
+    public void testEnterpriseFeature() {
+        LicensedFeature feature = LicensedFeature.momentary("family", "platinumFeature", ENTERPRISE);
```
Comment (ywangd) [inline x-pack/plugin/core/src/test/java/org/elasticsearch/license/XPackLicenseStateTests.java:234]:
```suggestion
        LicensedFeature feature = LicensedFeature.momentary("family", "enterpriseFeature", ENTERPRISE);
```

File: x-pack/plugin/core/src/test/java/org/elasticsearch/license/XPackLicenseStateTests.java [hunk 2]
```diff
@@ -201,246 +200,44 @@ public void testMonitoringAckNotBasicToBasic() {
...
-        assertThat(licenseState.checkFeature(XPackLicenseState.Feature.JDBC), is(false));
-    }
-
-    public void testSqlAckAnyToTrialOrPlatinum() {
-        assertAckMessages(XPackField.SQL, randomMode(), randomTrialOrPlatinumMode(), 0);
-    }
-
-    public void testSqlAckTrialOrPlatinumToNotTrialOrPlatinum() {
-        assertAckMessages(XPackField.SQL, randomTrialOrPlatinumMode(), randomBasicStandardOrGold(), 1);
-    }
```
Comment (ywangd) [inline x-pack/plugin/core/src/test/java/org/elasticsearch/license/XPackLicenseStateTests.java:375]: These test the acknowledgement messages between licenses for specific features. I think they are not covered by the new tests and should therefore be retained.

File: x-pack/plugin/core/src/test/java/org/elasticsearch/license/XPackLicenseStateTests.java [hunk 3]
```diff
@@ -201,246 +200,44 @@ public void testMonitoringAckNotBasicToBasic() {
...
-
-        assertFalse(state.checkFeature(XPackLicenseState.Feature.CCR));
-    }
-
-    public void testCcrAckAnyToTrialOrPlatinum() {
-        assertAckMessages(XPackField.CCR, randomMode(), randomTrialOrPlatinumMode(), 0);
-    }
-
-    public void testCcrAckTrialOrPlatinumToNotTrialOrPlatinum() {
-        assertAckMessages(XPackField.CCR, randomTrialOrPlatinumMode(), randomBasicStandardOrGold(), 1);
```
Comment (ywangd) [inline x-pack/plugin/core/src/test/java/org/elasticsearch/license/XPackLicenseStateTests.java:443]: Same as above.

File: x-pack/plugin/core/src/test/java/org/elasticsearch/license/XPackLicenseStateTests.java [hunk 4]
```diff
@@ -201,246 +200,44 @@ public void testMonitoringAckNotBasicToBasic() {
...
-
-    public void testWatcherInactive() {
-        assertAllowed(BASIC, false, s -> s.checkFeature(Feature.WATCHER), false);
-    }
-
-    public void testWatcherInactivePlatinumGoldTrial() throws Exception {
-        assertAllowed(TRIAL, false, s -> s.checkFeature(Feature.WATCHER), false);
-        assertAllowed(GOLD, false, s -> s.checkFeature(Feature.WATCHER), false);
-        assertAllowed(PLATINUM, false, s -> s.checkFeature(Feature.WATCHER), false);
-        assertAllowed(STANDARD, false, s -> s.checkFeature(Feature.WATCHER), false);
```
Comment (ywangd) [inline x-pack/plugin/core/src/test/java/org/elasticsearch/license/XPackLicenseStateTests.java:232]: The new tests do not cover the scenario of expired licenses. Do we need to add them with `LicensedFeature.momentaryLenient`? I know we plan to tighten the license expiration behaviour. So maybe it is not necessary?

**Review by ywangd — APPROVED** [review]
LGTM

I appreciate if you could address the comments. Thanks!

File: x-pack/plugin/core/src/test/java/org/elasticsearch/license/XPackLicenseStateTests.java (hunk 4 above)
Comment (rjernst) [reply x-pack/plugin/core/src/test/java/org/elasticsearch/license/XPackLicenseStateTests.java:232]: Sure, I can add a test for this, at least at a generic level.

**Comment by rjernst** [comment]
@elasticmachine run elasticsearch-ci/rest-compatibility

**Comment by rjernst** [comment]
@elasticmachine run elasticsearch-ci/part-2

**Pull request merged**